tms/watcher/watcher.py

```diff
@@ -0,0 +1,402 @@
+                    )
+                case _:
+                    set_job_status(JobInfoKey.JobStatus, job_status.value)
+        #
+        # OTHER
+        else:
+            raise UnknownJobEvent(f"not an important event: {job_event.type.name}")
+
+
+########################################################################################
+
+
+class EveryXSeconds:
+    """Keep track of durations."""
+
+    def __init__(self, seconds: float) -> None:
+        self.seconds = seconds
+        self._last_time = time.time()
+
+    def has_been_x_seconds(self) -> bool:
+        """Has it been at least `self.seconds` since last time?"""
+        diff = time.time() - self._last_time
+        yes = diff >= self.seconds
+        if yes:
+            self._last_time = time.time()
+            LOGGER.info(f"has been at least {self.seconds}s (actually {diff}s)")
+        return yes
+
+
+def is_file_past_modification_expiry(jel_fpath: Path) -> bool:
+    """Return whether the file was last modified longer than the expiry."""
+    diff = time.time() - jel_fpath.stat().st_mtime
+    yes = diff >= ENV.JOB_EVENT_LOG_MODIFICATION_EXPIRY
+    if yes:
+        LOGGER.warning(f"Job log file {jel_fpath} has not been updated in {diff}s")
+    return yes
+
+
+########################################################################################
+
+
+async def watch_job_event_log(jel_fpath: Path, ewms_rc: RestClient) -> None:
+    """Watch over one job event log file, containing multiple taskforces.
+
+    NOTE:
+        1. a taskforce is never split among multiple files, it uses only one
+        2. if this process crashes, the file will be re-read from the top;
+            so, there's no need to track progress.
+    """
+    LOGGER.info(f"This watcher will read {jel_fpath}")
+
+    cluster_info_dict: dict[str, ClusterInfo] = {}  # LARGE
+    time_tracker = EveryXSeconds(ENV.TMS_WATCHER_INTERVAL)
+    jel = htcondor.JobEventLog(str(jel_fpath))
+
+    while True:
+        # wait for job log to populate (more)
+        while not time_tracker.has_been_x_seconds():
+            await asyncio.sleep(1)
+
+        # get events -- exit when no more events, or took too long
+        got_new_events = False
+        LOGGER.info(f"Reading events from {jel_fpath}...")
+        for job_event in jel.events(stop_after=0):  # 0 -> only get currently available
+            got_new_events = True
+            if job_event.cluster not in cluster_info_dict:
+                cluster_info_dict[job_event.cluster] = ClusterInfo()
+            try:
+                cluster_info_dict[job_event.cluster].update_from_event(job_event)
+            except UnknownJobEvent as e:
+                LOGGER.debug(f"error: {e}")
+            await asyncio.sleep(0)  # since htcondor is not async
+            if time_tracker.has_been_x_seconds():
+                break
+
+        # endgame check
+        if not got_new_events:
+            if is_file_past_modification_expiry(jel_fpath):
```

Comment:
One thing I'm worried about is if there are still jobs sitting on the queue, but just haven't done anything recently.  Potentially things that are harder to match (or if we're having problems with the queue).

Maybe this is covered by the stopper, but I'm not sure.

Reply:
That's a good point. I can add a REST URL for that

Reply:
Given the current code, what happens in the case where we unlink the file, but there are still jobs on the queue in idle. Then the watcher restarts for some reason, and `cluster_infos` is reset.  Then one of those idle jobs runs and adds more info to `cluster_infos`, and this gets sent to the EWMS server.  Does this overwrite previous aggregations?  Do we lose a bunch of info then?

Reply:
I updated the endgame logic to require that all taskforces (as provided by ewms) are first seen in the job log. This still relies on an expiry, but it should remedy the issue of a cluster that is pending for a very long time in the queue

Reply:
This won't fully solve this issue, where one job takes a very long time; while the tms restarts. I can add some communication to when the watcher deems the job log finished

Reply:
see [361d638](https://github.com/Observation-Management-Service/ewms-task-management-service/pull/3/commits/361d63882ad1c53bfbfcc64327c51c314e63deb0)

Reply:
The easy answer is to remove any remaining jobs for that log file.  I'm not sure if you have a good way to do that though?